### PR TITLE
Annotate design document in preparation for update

### PR DESCRIPTION
**Pull request by spencerkimball**
@mjibson @petermattis @mrtracy @tschottdorf @arjunravinarayan @andreimatei @jordanlewis @jseldess @sploiselle 

All, you've been called to a higher purpose. Our design doc is in parlous condition. It's time to give it a miracle makeover. For those of you who feel the call of duty, we're going to organize a designdoc hackathon tomorrow at 1:30p to split design.md into areas of concern, parcel them out amongst ourselves, knock out the asbestos and water-stained drywall, and erect a modern, glass and steel structure in its place.

I've yet to delineate the missing sections (too late already tonight); we can spend some time on that tomorrow. But if we just get as far as dumping no-longer-applicable sections and cleaning up what remains, we'll be in significantly better shape.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9634)

### CODE REVIEW

**Comment by petermattis** [comment]
I'm happy to take on some of the work, though 1:30pm on Friday's is rough for me due to various shuttling of kids to activities.

**Comment by spencerkimball** [comment]
Why don't you comment in this PR for sections you'd like to take on.

File: docs/design.md [hunk 1]
```diff
@@ -101,12 +122,14 @@ Up to `F` failures can be tolerated, where the total number of replicas `N = 2F
...
 # Cockroach Client
 
+[Woah, big update required here!]
 In order to support diverse client usage, Cockroach clients connect to
 any node via HTTPS using protocol buffers or JSON. The connected node
 proxies involved client work including key lookups and write buffering.
 
 # Keys
 
+[Update referencing concepts in keys/constants.go required here]
```
Comment (petermattis) [inline docs/design.md]: I can tackle this section. Will probably rename it to `Internal Keys` to delineate it from `SQL`.

File: docs/design.md [hunk 2]
```diff
@@ -633,6 +663,7 @@ A really good reference on tuning Linux installations with RocksDB is
 
 # Range Metadata
 
+[Peter: did you actually have numbers for the size of meta2 records?]
```
Comment (petermattis) [inline docs/design.md]: Yes. Easy enough to recompute, though it was below 256 bytes and that upper bound still seems reasonable. I can take this section as well.

**Comment by spencerkimball** [comment]
+cc @knz

File: docs/design.md [hunk 3]
```diff
@@ -64,15 +74,25 @@ grained data on the level of entity groups.
...
 which handles the details of range addressing to provide the abstraction
 of a single, monolithic key value store. The distributed KV store
 communicates with any number of physical cockroach nodes. Each node
 contains one or more stores, one per physical device.
 
+![SQL]
+
+
+
+![Distributed SQL]
```
Comment (andreimatei) [inline docs/design.md]: I can take this

File: docs/design.md [hunk 4]
```diff
@@ -64,15 +74,25 @@ grained data on the level of entity groups.
...
 API*](#structured-data-api), which provides familiar relational concepts
+[Blend these structured data concepts into the SQL description]
 such as schemas, tables, columns, and indexes. The structured data API
 in turn depends on the [distributed key value store](#key-value-api),
 which handles the details of range addressing to provide the abstraction
 of a single, monolithic key value store. The distributed KV store
 communicates with any number of physical cockroach nodes. Each node
 contains one or more stores, one per physical device.
 
+![SQL]
```
Comment (andreimatei) [inline docs/design.md]: @knz was volunteered

**Comment by tschottdorf** [comment]
I'm happy to do whatever I'm being volunteered for.

File: docs/design.md [hunk 5]
```diff
@@ -64,15 +74,25 @@ grained data on the level of entity groups.
 
 Cockroach implements a layered architecture. The highest level of
 abstraction is the SQL layer (currently unspecified in this document).
+[FIX]
```
Comment (andreimatei) [inline docs/design.md]: I'll fix this

File: docs/design.md [hunk 6]
```diff
@@ -83,6 +103,7 @@ raft. The color coding shows associated range replicas.
 
 ![Ranges](media/ranges.png)
 
+[Update RoachNode concept]
```
Comment (spencerkimball) [inline docs/design.md]: I'll fix references to `RoachNode`.

File: docs/design.md [hunk 7]
```diff
@@ -101,12 +122,14 @@ Up to `F` failures can be tolerated, where the total number of replicas `N = 2F
 
 # Cockroach Client
```
Comment (spencerkimball) [inline docs/design.md]: I'll delete this whole section.

File: docs/design.md (hunk 1 above)
Comment (spencerkimball) [inline docs/design.md]: @tamird will do this section on keys.

**Comment by tschottdorf** [comment]
I'll get to the Linearizability section.

**Comment by arjunravinarayan** [comment]
I'll take Raft (+ Quiescing + Coalescing).

File: docs/design.md [hunk 8]
```diff
@@ -377,6 +401,7 @@ An exploration of retries with contention and abort times with abandoned
 transaction is
 [here](https://docs.google.com/document/d/1kBCu4sdGAnvLqpT-_2vaTbomNmX3_saayWEGYu1j7mQ/edit?usp=sharing).
 
+[Update (conceptually) s/table/records/g]
```
Comment (spencerkimball) [inline docs/design.md]: @tamird could you address this comment?

File: docs/design.md [hunk 9]
```diff
@@ -4,6 +4,9 @@ by Spencer Kimball from early 2014.
 
 # Overview
 
+[Spencer to edit overview. This stuff is so ripe it's starting to
+stink: "structured data? WTF is that?]
+
 Cockroach is a distributed key:value datastore (SQL and structured
```
Comment (knz) [inline docs/design.md]: I'm rewriting the SQL reference here.

File: docs/design.md [hunk 10]
```diff
@@ -14,6 +17,10 @@ disruption and **no manual intervention**. Cockroach nodes are
 symmetric; a design goal is **homogeneous deployment** (one binary) with
 minimal configuration.
 
+[Start at top of the new stack here: SQL, mention of distributed SQL,
```
Comment (knz) [inline docs/design.md]: Taking this over

File: docs/design.md (hunk 5 above)
Comment (knz) [inline docs/design.md]: No I will

File: docs/design.md [hunk 11]
```diff
@@ -1305,6 +1356,7 @@ write-optimized (HBase, Cassandra, SQLite3/LSM, CockroachDB).
 
 ## Architecture
 
+[Keep this diagram; update if necessary]
 CockroachDB implements a layered architecture, with various
 subdirectories implementing layers as appropriate. The highest level of
 abstraction is the [SQL layer][5], which depends
```
Comment (knz) [inline docs/design.md]: Rewriting this to clean up the SQL references

File: docs/design.md [hunk 12]
```diff
@@ -1331,6 +1383,7 @@ replicas.
 
 ## Client Architecture
 
+[Update this diagram]
```
Comment (knz) [inline docs/design.md]: Taking this over to explain SQL connections

**Comment by arjunravinarayan** [comment]
Review status: 0 of 1 files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[docs/design.md, line 812 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9634#-KSwEnI9ijiPcmpWzlHe:-KSwEnIAX5p11WMqWsUp:bu2hx09) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b112c0329f8faaf44e1024aeac61eecf03e660a/docs/design.md#L812)):*
> ```Markdown
> # Raft - Consistency of Range Replicas
> 
> [Raft experts review this nonsense below]
> ```

Taking this.

---

*[docs/design.md, line 878 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9634#-KSwErUBqOoDMHxY4chS:-KSwErUBqOoDMHxY4chT:bu2hx09) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b112c0329f8faaf44e1024aeac61eecf03e660a/docs/design.md#L878)):*
> ```Markdown
> ## Relationship to Raft leadership
> 
> [Is this all still accurate?]
> ```

Taking this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9634)*

File: docs/design.md [hunk 13]
```diff
@@ -609,6 +638,7 @@ for further details.
 
 # Node Storage
 
+[Updates necessary below]
```
Comment (BramGruneir) [inline docs/design.md]: I'll clean this section up a bit.  Remove the range tree.

File: docs/design.md [hunk 14]
```diff
@@ -660,6 +691,7 @@ is sparse, the successor key is defined as the next key which is present. The
 found using the same process. The *meta2* record identifies the range
 containing `key1`, which is again found the same way (see examples below).
 
+[Updates required below]
```
Comment (spencerkimball) [inline docs/design.md]: @tamird could you update here as well?

File: docs/design.md [hunk 15]
```diff
@@ -980,8 +1019,11 @@ spare capacity is chosen in the same datacenter and a special-case split
 is done which simply duplicates the data 1:1 and resets the range
 configuration metadata.
 
+[NEED a section on snapshots]
+
 # Range-Spanning Binary Tree
 
+[Remove this section I think.]
```
Comment (BramGruneir) [inline docs/design.md]: I'm on it.

File: docs/design.md [hunk 16]
```diff
@@ -903,6 +941,7 @@ expired, the command will be rejected by the replica.
 
 # Splitting / Merging Ranges
 
+[Accurate still? Needs review.]
```
Comment (BramGruneir) [inline docs/design.md]: Should we still discuss merges if we don't do them yet?

File: docs/design.md [hunk 17]
```diff
@@ -805,6 +839,7 @@ Future optimizations may include two-phase elections and quiescent ranges
 
 # Range Leases
 
+[Cleanup, add section on epoch-based range leases and motivation & link to range lease RFC]
```
Comment (spencerkimball) [inline docs/design.md]: @spencerkimball to do range leases

File: docs/design.md [hunk 18]
```diff
@@ -840,6 +875,7 @@ offset.
 
 ## Relationship to Raft leadership
 
+[Is this all still accurate?]
```
Comment (spencerkimball) [inline docs/design.md]: @arjunravinarayan to do this raft leadership section

File: docs/design.md [hunk 19]
```diff
@@ -1091,6 +1136,8 @@ The gossip protocol itself contains two primary components:
 
 # Node Accounting
 
+[Replace with description of MVCC, store-level aggregation, recorder,
```
Comment (spencerkimball) [inline docs/design.md]: @mrtracy would be great if you could replace `Node Accounting` section with an `Accounting`...

File: docs/design.md [hunk 20]
```diff
@@ -1225,6 +1275,7 @@ it discovers differences, it reconfigures ranges in the same way
 that it rebalances away from busy nodes, via special-case 1:1
 split to a duplicate range comprising the new configuration.
 
+[Rest of this document is tossable]
```
Comment (spencerkimball) [inline docs/design.md]: @spencerkimball to delete tail of document.

**Comment by petermattis** [comment]
If everyone is making edits to this doc, the merge might be painful. What is the plan?

**Comment by spencerkimball** [comment]
@petermattis just stay in your assigned section.

File: docs/design.md (hunk 19 above)
Comment (spencerkimball) [inline docs/design.md]: Sounds like @tschottdorf is handling this bit.

File: docs/design.md (hunk 15 above)
Comment (BramGruneir) [inline docs/design.md]: done #9643

File: docs/design.md (hunk 13 above)
Comment (BramGruneir) [inline docs/design.md]: done

**Pull request closed without merging**